nodes/stretch_ik_control.py

```diff
@@ -0,0 +1,1217 @@
+            Twist,
+            "/stretch/cmd_vel",
+            qos_profile=1,
+        )
+        self.arm_client_future = None
+        self.arm_client = ActionClient(
+            self.node,
+            FollowJointTrajectory,
+            "/stretch_controller/follow_joint_trajectory",
+            callback_group=MutuallyExclusiveCallbackGroup(),
+        )
+
+        self.initialized = True
+        return True
+
+    def __load_joint_limits(self, speed_profile: SpeedProfile) -> bool:
+        """
+        Loads the position and velocity limits for the robot's joints.
+
+        Note that we load a single limit for the arm, as opposed to separate limits
+        per telescoping joint. Also note that this function blocks until the service
+        to get joint limits is available.
+
+        Parameters
+        ----------
+        speed_profile: The speed profile to use to get max velocities and accelerations.
+
+        Returns
+        -------
+        bool: True if the joint limits were successfully loaded.
+        """
+        # Invoke the service to get joint limits. Note that this service
+        # does not actually return the joint limits. Rather, it publishes the joint
+        # limits to the /joint_limits topic.
+        get_joint_limits_client = self.node.create_client(
+            Trigger,
+            "get_joint_states",
+            # callback_group=MutuallyExclusiveCallbackGroup(),
+        )
+        self.node.get_logger().info("Waiting for the get joint states service...")
+        get_joint_limits_client.wait_for_service()  # TODO: This can block!
+        response = get_joint_limits_client.call(Trigger.Request())
+        if not response.success:
+            self.node.get_logger().error("Failed to get joint limits.")
+            return False
+        self.node.get_logger().info("Received the joint limits.")
+
+        # Get the robot parameters
+        robot_params = RobotParams().get_params()[1]
+        speed_profile_str = speed_profile.value
+        speed_profile_slow_str = SpeedProfile.SLOW.value
+        self.joint_pos_lim: Dict[Joint, Tuple[float, float]] = {}
+        self.joint_vel_abs_lim: Dict[Joint, Tuple[float, float]] = {}
+
+        # Get the velocity limits for the controllable joints
+        for joint_name in Joint:
+            # Get the max velocity
+            if joint_name == Joint.BASE_ROTATION:
+                # https://github.com/hello-robot/stretch_visual_servoing/blob/f99342/normalized_velocity_control.py#L21
+                max_abs_vel = 0.5  # rad/s
+                min_abs_vel = 0.05  # rad/s
+            elif joint_name == Joint.ARM_LIFT:
+                max_abs_vel = robot_params["lift"]["motion"][speed_profile_str]["vel_m"]
+                min_abs_vel = robot_params["lift"]["motion"][speed_profile_slow_str][
+                    "vel_m"
+                ]
+            elif joint_name == Joint.ARM_L0 or joint_name == Joint.COMBINED_ARM:
```

Comment:
Should we catch an error to make sure both of these dont happen

Reply:
There's no harm in both of these happening. In fact, it ends up being helpful to have both joints in the velocity limits dictionary, because the URDF uses `joint_arm_l0` but we command `joint_arm`, so it gives us more flexibility to apply velocity limits at both layers.